File: packages/lynx/host/android/app/src/main/java/com/yee94/openchamber/lynx/OpenChamberLynxVirtualAsset.kt
package com.yee94.openchamber.lynx

/** openchamber-asset://v/{assetId} resolver hooks. */
object OpenChamberLynxVirtualAsset {
    const val SCHEME = "openchamber-asset"
    const val MIME_MAX_LENGTH = 128

    fun url(assetId: String): String? {
        if (assetId.length !in 8..80) return null
        if (!assetId.all { it.isLetterOrDigit() || it == '_' || it == '-' }) return null
        return "$SCHEME://v/$assetId"
    }

    fun normalizeMime(mime: String): String? {
        val normalized = mime.trim().lowercase()
        if (normalized.isEmpty() || normalized.length > MIME_MAX_LENGTH) return null
        if (!normalized.startsWith("image/")) return null
        return normalized
    }
}

File: packages/lynx/host/ios/OpenChamberLynxVirtualAsset.swift
import Foundation

/// `openchamber-asset://v/{assetId}` hooks — Cap OpenChamberVirtualAsset spirit.
/// Register a WKURLSchemeHandler / Lynx resource provider when the SDK is linked.
enum OpenChamberLynxVirtualAsset {
  static let scheme = "openchamber-asset"
  static let mimeMaxLength = 128

  static func url(assetId: String) -> URL? {
    var allowed = CharacterSet.alphanumerics
    allowed.insert(charactersIn: "_-")
    guard assetId.count >= 8, assetId.count <= 80,
          assetId.unicodeScalars.allSatisfy({ allowed.contains($0) }) else {
      return nil
    }
    return URL(string: "\(scheme)://v/\(assetId)")
  }

  static func normalizeMime(_ mime: String) -> String? {
    let normalized = mime.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard !normalized.isEmpty, normalized.count <= mimeMaxLength else { return nil }
    guard normalized.hasPrefix("image/") else { return nil }
    return normalized
  }
}
